{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9f2b805e",
   "metadata": {
    "id": "9f2b805e"
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "46cdd0f1",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "46cdd0f1",
    "outputId": "716ab9e2-f8c9-41d7-b1de-ba9a9edfde3b"
   },
   "outputs": [],
   "source": [
    "#!pip install -q --no-warn-conflicts malariagen_data -U"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "571acbfe",
   "metadata": {},
   "source": [
    "This file accesses MalariaGen Pf7 data QC pass data at the variant level to calculate site missingness across all variants. It accesses Pf7 data stored via the cloud, which requires registration on the malariagen website and subsequent gcloud logins during each session. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d22e67b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gcloud auth application-default login"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "efc51f37",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 409
    },
    "id": "efc51f37",
    "outputId": "4ff39d12-1c87-4954-c862-bbec54d75d23"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import dask\n",
    "import dask.array as da\n",
    "from dask.diagnostics.progress import ProgressBar\n",
    "import allel\n",
    "# silence some warnings\n",
    "dask.config.set(**{'array.slicing.split_large_chunks': False})\n",
    "import malariagen_data\n",
    "import collections\n",
    "import pandas as pd\n",
    "import xarray as xr\n",
    "\n",
    "pf7 = malariagen_data.Pf7()\n",
    "pf7_metadata = pf7.sample_metadata()\n",
    "#pf7_metadata = pd.read_csv('Pf7_samples.txt', sep = '\\t')\n",
    "variant_dataset = pf7.variant_calls()\n",
    "genome_features0 = pf7.genome_features()\n",
    "# EXTENDED VARIANT DATA -- variant_ANN_Annotation_Impact\n",
    "extended_variant_dataset = pf7.variant_calls(extended=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8dc6126c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PF3D7_0829600 13\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[0.8448275862068966]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def pass_rate(stage):\n",
    "  # read in stage-specific gene set\n",
    "  #fname = stage + '.txt'\n",
    "  #gene_set = np.loadtxt(fname, dtype = str)\n",
    "  gene_set = stage\n",
    "  print(stage, len(gene_set))\n",
    "  # mark which genes are in gene set from BOOL\n",
    "  in_set = [(i in gene_set) for i in genome_features0['ID']]\n",
    "  # subset gene set features from list of all 3D7 genome features\n",
    "  genome_features = genome_features0[in_set]\n",
    "  contigs = genome_features['contig']\n",
    "  start = genome_features['start']\n",
    "  end = genome_features['end']\n",
    "\n",
    "  # APPROACH: PREPARE ALL MASKS AND COMBINE THEM TOGETHER AT END\n",
    "  # load in filter pass flags\n",
    "  fpass = variant_dataset['variant_filter_pass'].data\n",
    "  # variant pass dataset --> keep SNPs only (same for all)\n",
    "  # first: snp mask (filter array for SNPs only)\n",
    "  snp_mask = variant_dataset['variant_is_snp'].data\n",
    "\n",
    "  # coding mask: ensure variant falls in coding region\n",
    "  coding_mask = variant_dataset['variant_CDS'].data\n",
    "\n",
    "  pass_rates = []\n",
    "  # iterate through all gene coordinates pulled from reference genome\n",
    "  for i, j, k in zip(contigs, start, end):\n",
    "    # mask those on same contig\n",
    "    contig_mask = variant_dataset['variant_chrom'].data == i\n",
    "    # mask for position\n",
    "    pos_mask = da.isin(variant_dataset['variant_position'].data, np.arange(j, k))\n",
    "    # combine masks\n",
    "    gene_mask = da.logical_and(contig_mask, pos_mask)\n",
    "    gene_mask_coding = da.logical_and(gene_mask, coding_mask)\n",
    "    mask_final = da.logical_and(snp_mask, gene_mask_coding)\n",
    "    fpass_gene = fpass[mask_final.compute()]\n",
    "    pass_rate0 = da.sum(fpass_gene).compute()/fpass_gene.size\n",
    "    pass_rates.append(pass_rate0)\n",
    "    #print(pass_rate0)\n",
    "  return(pass_rates)\n",
    "\n",
    "pass_rate('PF3D7_0829600')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2daf849",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in gene sets, e.g. sets of genes expressed / detectable in assay with preliminary breadth labels \n",
    "# these labels will be changes later (after analysis)\n",
    "stages = ['gene_sets_assay/' + str(i) for i in range(1,7)]\n",
    "\n",
    "# calculate pass rates for each gene set loaded in \n",
    "rates_all = []\n",
    "for i in stages:\n",
    "  rates_all.append(pass_rate(i))\n",
    "\n",
    "rates_all[1]\n",
    "# save all to csv \n",
    "rdf = pd.DataFrame(rates_all).T\n",
    "rdf.columns = stages\n",
    "rdf.head()\n",
    "rdf.to_csv('missing_by_life_breadth.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "06a5e4b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n"
     ]
    }
   ],
   "source": [
    "!grep -v -F -f  gene_sets_all/all.txt props_adj_breadth.txt | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fea95fbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "rates_all = pass_rate('gene_sets_all/all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "LCeO88fqS39X",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 363
    },
    "id": "LCeO88fqS39X",
    "outputId": "e99d9db2-fc04-4f2c-b9a6-6faafd495431"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.821739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.783599</td>\n",
       "      <td>0.826748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.680769</td>\n",
       "      <td>0.836914</td>\n",
       "      <td>0.830470</td>\n",
       "      <td>0.657718</td>\n",
       "      <td>0.814491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>0.734072</td>\n",
       "      <td>0.740634</td>\n",
       "      <td>0.820652</td>\n",
       "      <td>0.784768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.767810</td>\n",
       "      <td>0.864286</td>\n",
       "      <td>0.849462</td>\n",
       "      <td>0.798030</td>\n",
       "      <td>0.836263</td>\n",
       "      <td>0.823819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.844311</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.805310</td>\n",
       "      <td>0.805556</td>\n",
       "      <td>0.836237</td>\n",
       "      <td>0.775079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.792453</td>\n",
       "      <td>0.808824</td>\n",
       "      <td>0.753425</td>\n",
       "      <td>0.801158</td>\n",
       "      <td>0.837017</td>\n",
       "      <td>0.789062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.684211</td>\n",
       "      <td>0.757353</td>\n",
       "      <td>0.813830</td>\n",
       "      <td>0.829431</td>\n",
       "      <td>0.835667</td>\n",
       "      <td>0.719178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.786325</td>\n",
       "      <td>0.818966</td>\n",
       "      <td>0.619718</td>\n",
       "      <td>0.801762</td>\n",
       "      <td>0.842844</td>\n",
       "      <td>0.769231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.036269</td>\n",
       "      <td>0.861789</td>\n",
       "      <td>0.751880</td>\n",
       "      <td>0.829604</td>\n",
       "      <td>0.781457</td>\n",
       "      <td>0.853211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.764228</td>\n",
       "      <td>0.693548</td>\n",
       "      <td>0.762376</td>\n",
       "      <td>0.796316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          1         2         3         4         5         6\n",
       "0  0.000000  0.000000  0.821739  0.000000  0.783599  0.826748\n",
       "1  0.000000  0.680769  0.836914  0.830470  0.657718  0.814491\n",
       "2  0.000000  0.794872  0.734072  0.740634  0.820652  0.784768\n",
       "3  0.767810  0.864286  0.849462  0.798030  0.836263  0.823819\n",
       "4  0.844311  0.833333  0.805310  0.805556  0.836237  0.775079\n",
       "5  0.792453  0.808824  0.753425  0.801158  0.837017  0.789062\n",
       "6  0.684211  0.757353  0.813830  0.829431  0.835667  0.719178\n",
       "7  0.786325  0.818966  0.619718  0.801762  0.842844  0.769231\n",
       "8  0.036269  0.861789  0.751880  0.829604  0.781457  0.853211\n",
       "9  0.000000  0.000000  0.764228  0.693548  0.762376  0.796316"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# runIF new session for analysis\n",
    "import pandas as pd\n",
    "rdf = pd.read_csv('missing_by_life_breadth.csv', index_col=0)\n",
    "#stages = ['ookinete', 'gametocyte', 'sporozoite', 'schizont', 'trophozoite', 'ring']\n",
    "stages = range(1,7)\n",
    "rdf.columns = stages\n",
    "rdf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7e91400",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine with DF containing site counts \n",
    "props = pd.read_csv('PlasmoDB-61_Pfalciparum3D7_AnnotatedCDSs_SynAndNonsynSiteCount.txt', sep = '\\t')\n",
    "rdfm = pd.melt(rdf)\n",
    "rdfm = rdfm[~np.isnan(rdfm['value'])] # filter out junk cells from melting procedure \n",
    "rdfm['ID'] = genes['value']\n",
    "props = props.merge(rdfm, left_on='GENE', right_on='ID')\n",
    "props['coding_length_adj'] = props['TOTAL_CODING_LENGTH'] * props['value']\n",
    "props.to_csv('props_adj_breadth.txt', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "9b9fab73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>TRANS</th>\n",
       "      <th>NAME</th>\n",
       "      <th>NS</th>\n",
       "      <th>SYN</th>\n",
       "      <th>FFD</th>\n",
       "      <th>TOTAL_CODING_LENGTH</th>\n",
       "      <th>PROP_NS</th>\n",
       "      <th>PROP_SYN</th>\n",
       "      <th>PROP_FFD</th>\n",
       "      <th>STOP_CODONS</th>\n",
       "      <th>FULL_LENGTH</th>\n",
       "      <th>COORD</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "      <th>ID</th>\n",
       "      <th>coding_length_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PF3D7_1137800</td>\n",
       "      <td>PF3D7_1137800.1</td>\n",
       "      <td>sporozoite surface protein essential for liver...</td>\n",
       "      <td>342.833333</td>\n",
       "      <td>113.166667</td>\n",
       "      <td>62</td>\n",
       "      <td>456</td>\n",
       "      <td>0.751827</td>\n",
       "      <td>0.248173</td>\n",
       "      <td>0.135965</td>\n",
       "      <td>1</td>\n",
       "      <td>627</td>\n",
       "      <td>location=Pf3D7_11_v3:1480900-1481526(-)</td>\n",
       "      <td>sporozoite</td>\n",
       "      <td>0.844828</td>\n",
       "      <td>PF3D7_1137800</td>\n",
       "      <td>385.241379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PF3D7_0932200</td>\n",
       "      <td>PF3D7_0932200.1</td>\n",
       "      <td>profilin</td>\n",
       "      <td>403.333333</td>\n",
       "      <td>112.666667</td>\n",
       "      <td>65</td>\n",
       "      <td>516</td>\n",
       "      <td>0.781654</td>\n",
       "      <td>0.218346</td>\n",
       "      <td>0.125969</td>\n",
       "      <td>1</td>\n",
       "      <td>1365</td>\n",
       "      <td>location=Pf3D7_09_v3:1287008-1288372(+)</td>\n",
       "      <td>schizont</td>\n",
       "      <td>0.801178</td>\n",
       "      <td>PF3D7_0932200</td>\n",
       "      <td>413.407740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PF3D7_0901500</td>\n",
       "      <td>PF3D7_0901500.1</td>\n",
       "      <td>rifin</td>\n",
       "      <td>853.166667</td>\n",
       "      <td>244.833333</td>\n",
       "      <td>137</td>\n",
       "      <td>1098</td>\n",
       "      <td>0.777019</td>\n",
       "      <td>0.222981</td>\n",
       "      <td>0.124772</td>\n",
       "      <td>1</td>\n",
       "      <td>1338</td>\n",
       "      <td>location=Pf3D7_09_v3:74409-75746(+)</td>\n",
       "      <td>sporozoite</td>\n",
       "      <td>0.839071</td>\n",
       "      <td>PF3D7_0901500</td>\n",
       "      <td>921.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PF3D7_1116900</td>\n",
       "      <td>PF3D7_1116900.1</td>\n",
       "      <td>conserved protein, unknown function</td>\n",
       "      <td>499.000000</td>\n",
       "      <td>140.000000</td>\n",
       "      <td>79</td>\n",
       "      <td>639</td>\n",
       "      <td>0.780908</td>\n",
       "      <td>0.219092</td>\n",
       "      <td>0.123631</td>\n",
       "      <td>1</td>\n",
       "      <td>1438</td>\n",
       "      <td>location=Pf3D7_11_v3:644100-645537(-)</td>\n",
       "      <td>sporozoite</td>\n",
       "      <td>0.754098</td>\n",
       "      <td>PF3D7_1116900</td>\n",
       "      <td>481.868852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PF3D7_1245800</td>\n",
       "      <td>PF3D7_1245800.1</td>\n",
       "      <td>epsin-like protein, putative</td>\n",
       "      <td>1008.833333</td>\n",
       "      <td>287.166667</td>\n",
       "      <td>152</td>\n",
       "      <td>1296</td>\n",
       "      <td>0.778421</td>\n",
       "      <td>0.221579</td>\n",
       "      <td>0.117284</td>\n",
       "      <td>1</td>\n",
       "      <td>1616</td>\n",
       "      <td>location=Pf3D7_12_v3:1908149-1909764(+)</td>\n",
       "      <td>gametocyte</td>\n",
       "      <td>0.842589</td>\n",
       "      <td>PF3D7_1245800</td>\n",
       "      <td>1091.995587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            GENE            TRANS  \\\n",
       "0  PF3D7_1137800  PF3D7_1137800.1   \n",
       "1  PF3D7_0932200  PF3D7_0932200.1   \n",
       "2  PF3D7_0901500  PF3D7_0901500.1   \n",
       "3  PF3D7_1116900  PF3D7_1116900.1   \n",
       "4  PF3D7_1245800  PF3D7_1245800.1   \n",
       "\n",
       "                                                NAME           NS         SYN  \\\n",
       "0  sporozoite surface protein essential for liver...   342.833333  113.166667   \n",
       "1                                           profilin   403.333333  112.666667   \n",
       "2                                              rifin   853.166667  244.833333   \n",
       "3                conserved protein, unknown function   499.000000  140.000000   \n",
       "4                       epsin-like protein, putative  1008.833333  287.166667   \n",
       "\n",
       "   FFD  TOTAL_CODING_LENGTH   PROP_NS  PROP_SYN  PROP_FFD  STOP_CODONS  \\\n",
       "0   62                  456  0.751827  0.248173  0.135965            1   \n",
       "1   65                  516  0.781654  0.218346  0.125969            1   \n",
       "2  137                 1098  0.777019  0.222981  0.124772            1   \n",
       "3   79                  639  0.780908  0.219092  0.123631            1   \n",
       "4  152                 1296  0.778421  0.221579  0.117284            1   \n",
       "\n",
       "   FULL_LENGTH                                    COORD    variable     value  \\\n",
       "0          627  location=Pf3D7_11_v3:1480900-1481526(-)  sporozoite  0.844828   \n",
       "1         1365  location=Pf3D7_09_v3:1287008-1288372(+)    schizont  0.801178   \n",
       "2         1338      location=Pf3D7_09_v3:74409-75746(+)  sporozoite  0.839071   \n",
       "3         1438    location=Pf3D7_11_v3:644100-645537(-)  sporozoite  0.754098   \n",
       "4         1616  location=Pf3D7_12_v3:1908149-1909764(+)  gametocyte  0.842589   \n",
       "\n",
       "              ID  coding_length_adj  \n",
       "0  PF3D7_1137800         385.241379  \n",
       "1  PF3D7_0932200         413.407740  \n",
       "2  PF3D7_0901500         921.300000  \n",
       "3  PF3D7_1116900         481.868852  \n",
       "4  PF3D7_1245800        1091.995587  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "props = pd.read_csv(\"props_adj.txt\", sep='\\t')\n",
    "props.head()"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
